File: src/main/java/com/leijendary/spring/iamtemplate/config/properties/VerificationProperties.java
package com.leijendary.spring.iamtemplate.config.properties;

import lombok.Data;
import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties(prefix = "verification")
@Data
public class VerificationProperties {

    // In minutes
    private int expiry;
    private Register register;
    private ResetPassword resetPassword;

    @Data
    public static class Register {

        private String url;
    }

    @Data
    public static class ResetPassword {

        private String url;
    }
}

File: src/main/kotlin/com/leijendary/spring/template/iam/core/config/properties/VerificationProperties.kt
package com.leijendary.spring.template.iam.core.config.properties

import com.leijendary.spring.template.iam.core.util.RequestContext.now
import org.springframework.boot.context.properties.ConfigurationProperties
import java.time.Duration
import java.time.OffsetDateTime

@ConfigurationProperties(prefix = "verification")
class VerificationProperties {
    var expiry: Duration = Duration.ofHours(2)
    var timeout: Duration = Duration.ofSeconds(60)
    var register = Config()
    var password = Password()
    var email = Config()

    open inner class Config {
        lateinit var template: String
    }

    inner class Password {
        var nominate = PasswordNominate()
        var reset = Config()
    }

    inner class PasswordNominate : Config() {
        lateinit var url: String
    }

    fun computeExpiration(): OffsetDateTime = now.plus(expiry)

    fun computeTimeout(): OffsetDateTime = now.minusSeconds(timeout.toSeconds())
}
